gensim/models/word2vec.py

```diff
@@ -1144,9 +1150,43 @@ def __iter__(self):
                         i += self.max_sentence_length
 
 
+def _scan_vocab_worker(stream, progress_queue, max_vocab_size=None, trim_rule=None):
+    """Do an initial scan of all words appearing in stream.
+
+    Note: This function can not be Word2VecVocab's method because
+    of multiprocessing synchronization specifics in Python.
+    """
+    min_reduce = 1
+    vocab = defaultdict(int)
+    checked_string_types = 0
+    sentence_no = -1
+    total_words = 0
+    for sentence_no, sentence in enumerate(stream):
+        if not checked_string_types:
+            if isinstance(sentence, string_types):
+                log_msg = "Each 'sentences' item should be a list of words (usually unicode strings). " \
+                          "First item here is instead plain %s." % type(sentence)
+                progress_queue.put(log_msg)
+
+            checked_string_types += 1
+
+        for word in sentence:
+            vocab[word] += 1
+
+        if max_vocab_size and len(vocab) > max_vocab_size:
+            utils.prune_vocab(vocab, min_reduce, trim_rule=trim_rule)
+            min_reduce += 1
+
+        total_words += len(sentence)
+
+    progress_queue.put((total_words, sentence_no + 1))
+    progress_queue.put(None)
+    return vocab
```

Comment:
This will try to pickle the entire result (huge!) and send it back to master. Can you think of a way to use shared memory instead?

That should improve both memory (in a big way) as well as maybe performance (some extra locking, but less data sending).

Reply:
Sorry, missed this notification :)

My thoughts: you mean processes, not threads, right? Shrinking could be tricker, I agree. There'd have to be a lock around that (workers pause while shrinking under way -- hopefully not that frequent).

Disabling pruning with multistream is also an elegant solution :-)

Reply:
@piskvorky 

Pausing all workers while doing shrinking -- sounds very slow. Why are we sure that shrinking is performed not that frequent?

Do you want me to implement this kind of logic and benchmark it? Then decide which logic we will merge into develop branch?

Reply:
I would hope so! It's a costly operation. That's also why we do it with a growing threshold, instead of always clipping against a constant, e.g. `remove if count <= 5`.

An experiment would be worthwhile IMO. Not necessarily a full-fledged merge-able implementation, but some insight into whether we can get away with sharing a single structure among all workers. It would certainly save us lots of memory.

Reply:
Okay, I'll do that :wink:

Reply:
@piskvorky I've tried two approaches with `manager.dict()`.

1. Replace local dicts in each worker with `manager.dict()` and perform `vocab[word] += 1` on it. Also, in this case, I don't return `vocab` from each worker, but use shared dict at the end of scanning. This is the approach we were talking about above. The code is here https://gist.github.com/persiyanov/76560df460e8b80d4c6c6124452518c0

2. The second approach is still using local `dict()`-s in each worker and not returning them from worker's functions, but perform `manager.dict().update(local_vocab)`. The code is here https://gist.github.com/persiyanov/8537a81fe9f579aa5470ff27287607bf


Unfortunately, both approaches run very slowly and definitely have worse performance than the approach in current PR.

• 1st way (with updating shared vocab immediately with each word) runs `> 15 minutes` for all runs (`workers=2,3,4,5,8,10`), so I didn't even wait till finish.
• 2st way has the following timings:

```
2018-06-24 15:17:05,682 : INFO : Model = word2vec	Workers = 2	Vocab time = 291.13 secs
2018-06-24 15:21:21,838 : INFO : Model = word2vec	Workers = 3	Vocab time = 255.81 secs
2018-06-24 15:25:17,943 : INFO : Model = word2vec	Workers = 4	Vocab time = 235.76 secs
2018-06-24 15:29:01,697 : INFO : Model = word2vec	Workers = 5	Vocab time = 223.41 secs
2018-06-24 15:32:36,754 : INFO : Model = word2vec	Workers = 8	Vocab time = 214.70 secs
2018-06-24 15:35:58,231 : INFO : Model = word2vec	Workers = 10	Vocab time = 201.11 secs
```

It's clearly worse than what we have.

Reply:
Alternate idea: discard `mac_vocab_size`-style trimming entirely, figure out a way to do a precise count, by spilling extras to disk temp files. Strawman implementation: wherever the old code would've trimmed, instead dump the counts that would've been discarded to a key-sorted temp file. At the end, merge all the temp files to a final combined sorted count temp file – and while doing that, retain a histogram of word-frequencies. From that histogram, the portion of the structure in heap can calculate exactly how many words must be dropped to exactly match the specified cap, and all surviving words' counts will be precisely accurate. 

I doubt the performance hit would be very much, and it's fairly parallelizable (there's just more temp files to include in the final merge), the final counts are accurate, and there are fewer gotchas that need documenting because in the end it "just works" as one might expect.

Maybe, using a LevelDB-like log-structured-merge DB might achieve a similar benefit at the same or close performance.

Reply:
I forget to note I had commented `max_vocab_size`-trimming lines of code in both approaches above. So it's very slow even without trimming at all (and then without its synchronization overhead)

Reply:
Hm, strange. Must be due to the locking, right?

@persiyanov Any way to turn locking off (since we don't care about accurate counts, especially in the high-frequency tokens where collisions are more likely)?

Reply:
@piskvorky Yes, I bet this happens because of locks.

> Any way to turn locking off?

I can only see dirty ways to do that, e.g. make nogil Cython analogue for `defaultdict(int)`.

Reply:
I see. Have we exhausted ideas for memory sharing? Anything else we could try easily?

If not, are we finished with parallelized dict-building?

Reply:
@piskvorky done, see the updated description

Reply:
I made some minor language edits, looks good :) Thanks!